api/src/opentrons/protocol_api/validation.py

```diff
@@ -180,3 +196,57 @@ def ensure_valid_labware_offset_vector(
     if not all(isinstance(v, (float, int)) for v in offsets):
         raise TypeError("Offset values should be a number (int or float).")
     return offsets
+
+
+class WellTarget(NamedTuple):
+    """A movement target that is a well."""
+
+    well: Well
+    location: Optional[Location]
+
+
+class PointTarget(NamedTuple):
+    """A movement to coordinates"""
+
+    location: Location
+    in_place: bool = False
+
+
+class NoLocationError(ValueError):
+    """Error representing that no location was supplied."""
+
+
+class LocationTypeError(TypeError):
+    """Error representing that the location supplied is of different expected type."""
+
+
+def validate_location(
+    location: Union[Location, Well, None], last_location: Optional[Location]
+) -> Union[WellTarget, PointTarget]:
+    """Validate and return if should use a WellTarget or a PointTarget.
+    Args:
+        location: The input location.
+        last_location: The cached last location.
+    Raises:
+        NoLocationError: The is no input location and no cached loaction.
+        LocationTypeError: The location supplied is of unexpected type.
+    """
+    from .labware import Well
+
+    if location is None and last_location is None:
+        raise NoLocationError()
+
+    if isinstance(location, Well):
+        return WellTarget(well=location, location=None)
+    elif isinstance(location, Location):
+        _, well = location.labware.get_parent_labware_and_well()
+        if well is None:
+            return PointTarget(location=location, in_place=False)
+        return WellTarget(well=well, location=location)
+    elif last_location and location is None:
+        _, well = last_location.labware.get_parent_labware_and_well()
+        if well is None:
+            return PointTarget(location=last_location, in_place=True)
+        return WellTarget(well=well, location=last_location)
+
+    raise LocationTypeError()
```

Comment:
Given that we now have a good unit test suite for this function, we can take a refactor pass to combine common logic and improve readability

```suggestion
    target_location = location or last_location

    if target_location is None:
        raise NoLocationError()

    if not isinstance(target_location, (Location, Well)):
        raise LocationTypeError()

    if isinstance(target_location, Well):
        return WellTarget(well=target_location, location=None)

    _, well = target_location.labware.get_parent_labware_and_well()
    in_place = target_location is last_location

    return (
        WellTarget(well=well, location=target_location)
        if well is not None
        else PointTarget(location=target_location, in_place=in_place)
    )
```